pkg/datastore/phpapi/phpapi.go

```diff
@@ -48,59 +50,76 @@ func (p *PHPAPI) GetClusters(colorCodes []string) ([]types.SubmarinerCluster, er
+	var clusters []types.SubmarinerCluster
+	if err = json.Unmarshal(clustersRaw, &clusters); err != nil {
+		return nil, fmt.Errorf("error unmarshalling JSON %s: %v", string(clustersRaw[:]), err)
 	}
 	return clusters, nil
 }
+
 func (p *PHPAPI) GetCluster(clusterID string) (types.SubmarinerCluster, error) {
 	requestURL := fmt.Sprintf("%s://%s/clusters.php?plurality=false&identifier=%s&cluster_id=%s", p.Proto, p.Server, p.APIToken, clusterID)
 	klog.V(8).Infof("request url: %s", requestURL)
 	clustersGetter, err := http.Get(requestURL)
 	if err != nil {
-		klog.Errorf("Encountered error while trying to retrieve clusters from the apiserver")
-		return types.SubmarinerCluster{}, err
+		return types.SubmarinerCluster{}, fmt.Errorf("error retrieving clusters from %s: %v", requestURL, err)
 	}
+
 	defer clustersGetter.Body.Close()
 	clustersRaw, err := ioutil.ReadAll(clustersGetter.Body)
+	if err != nil {
+		return types.SubmarinerCluster{}, fmt.Errorf("error reading response from %s: %v", requestURL, err)
+	}
+
 	klog.V(8).Infof("response body: %v", string(clustersRaw[:]))
+
 	var cluster types.SubmarinerCluster
-	json.Unmarshal(clustersRaw, &cluster)
+	if err = json.Unmarshal(clustersRaw, &cluster); err != nil {
+		return types.SubmarinerCluster{}, fmt.Errorf("error unmarshalling JSON %s: %v", string(clustersRaw[:]), err)
+	}
 
 	return cluster, nil
 }
+
 func (p *PHPAPI) GetEndpoints(clusterID string) ([]types.SubmarinerEndpoint, error) {
 	requestURL := fmt.Sprintf("%s://%s/endpoints.php?plurality=true&identifier=%s&cluster_id=%s", p.Proto, p.Server, p.APIToken, clusterID)
 	endpointsGetter, err := http.Get(requestURL)
 	if err != nil {
-		klog.Errorf("encountered error while trying to retrieve endpoints from the apiserver: %v", err)
-		return nil, err
+		return nil, fmt.Errorf("error retrieving endpoints from %s: %v", requestURL, err)
 	}
+
 	defer endpointsGetter.Body.Close()
 	endpointsRaw, err := ioutil.ReadAll(endpointsGetter.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response from %s: %v", requestURL, err)
+	}
+
 	klog.V(8).Infof("response body: %v", string(endpointsRaw[:]))
+
 	var endpoints []types.SubmarinerEndpoint
-	err = json.Unmarshal(endpointsRaw, &endpoints)
-	if err != nil {
-		klog.Errorf("error while unmarshaling JSON from endpoints")
-		return nil, err
+	if err = json.Unmarshal(endpointsRaw, &endpoints); err != nil {
+		return nil, fmt.Errorf("error unmarshalling JSON %s: %v", string(endpointsRaw[:]), err)
 	}
 	return endpoints, nil
 }
+
 func (p *PHPAPI) GetEndpoint(clusterID string, cableName string) (types.SubmarinerEndpoint, error) {
-	klog.Errorf("GetEndpoint not implemented")
-	return types.SubmarinerEndpoint{}, nil
+	return types.SubmarinerEndpoint{}, fmt.Errorf("GetEndpoint is not implemented")
 }
+
 func (p *PHPAPI) WatchClusters(ctx context.Context, selfClusterID string, colorCodes []string, onChange func(cluster types.SubmarinerCluster, deleted bool) error) error {
```

Comment:
This no longer returns errors, AFAICT.

Reply:
It didn't before - the only return is nil at the end of the method.

Reply:
Right, what I was really thinking was that, as part of the error-handling cleanup, it would make sense to change the method declaration to indicate that it doesn’t return anything.

Reply:
yeah but it's part of the interface contract although, as it turns out, neither implementation (the other is kubernetes) returns an error. Of course when defining an interface it's safer to allow for error as you really can't anticipate the behavior of every implementation. In this case, performing the watch is assumed to be an async process so not much reason to return error.  I'll push another PR for that.

Reply:
@skitt I'd like to progress this PR soon as it cleans up the remaining golangci-lint warnings 😄 Are you OK the way it is?

Reply:
Yes, let’s get this merged.